### ISSUE TITLE
[RFC] Need to be able to compile only rustpython's virtual machine to wasm

### ISSUE THREAD

**Issue by Joetib** [RustPython/RustPython#1760]
### RustPython in the web can be one of it's best use cases
The python community have been pushing to have python run in the browser.
Many such implementations have been brython, skulpt, Batavia etc.
Likened with batavia's approach to run ```python bytecode ``` in the browser, I have been wondering if rustpytgon can ship only it's virtual machine to do similar task

## Detailed Explanation
Rustpython's  wasm module currently is about 4.5mb. thus size is too big to be inculcated in most projects. 
However I believe if we could remove the parser, lexer and other compiler related stuff and compile only the virtual machine, the size will drastically reduce.
This approach will also require that apps get compiled on their maybe machines before the bytecode is shipped.

## Drawbacks, Rationale, and Alternative

# Rationale:
 It will take a long time for RustPython to get to compete with cpython natively(where cpython works well already) but the web is one place where competition is less. If this works, we can have web apps fully written in python(little to no javascript) 
Be able to run already existing software on the web.
This can also lead to the revolution of building front end livraries. Like implementing python alternatives to react, angular, Vue etc. 
Currently, the beeware team have been working hard to have the Toga(beewae's gui library) working in browsers but are limited by the size of Batavia.
I may certain everyone loved Anvil.works for their idea of creating web apps completely in python. 

# Possible Drawback: 
Since apps will need to be compiled before shipped and most people use cpython, rustpython's vm must be able to run cpython generated bytecode 
This may not even be an issue but I raise this because I'm unaware of the bytecode implementation.

**Comment by coolreader18** [comment]
This is a really nice idea; we've experimented with excluding the parser and compiler for native executables of rustpython, but not yet for wasm. We don't currently run CPython bytecode; we have our own bytecode format and I think it would be less effort to have people install the rustpython compiler as part of whatever toolchain this would be than it would be to convert CPython's bytecode to RustPython's. Also, I could see bundling 2 different versions of rustpython into the `rustpython_wasm` npm package; one with the compiler (`import "rustpython_wasm"`) and one with only the vm (`import "rustpython_wasm/vm-only"`).

**Comment by Joetib** [comment]
Well I think the proposed fix with regards to cpython bytecode (thus having to install RustPython for the toolchain) is great and won't be difficult implementing.

If excluding the parser is capable of generating smaller wasm file probably just kilobytes, then if the size is still too large, a little tree shaking can be done especially on parts of the standard library that won't be needed on the web.

**Comment by Joetib** [comment]
I would love to ask if it will be a simple task to compile only the vm to webassembly as I am unsure of how it depends on the other modules. And if possible, suggestive steps to make that work ?

**Comment by coolreader18** [comment]
Yep, you just have to disable the default features (which right now is `rustpython-parser` and `rustpython-compiler`), and then the VM won't depend on the parser and compiler crates:

```toml
[features]
parser-compiler = [
    "rustpython-parser", "rustpython-compiler",
    "rustpython-vm/rustpython-parser", "rustpython-vm/rustpython-compiler"
]
default = ["parser-compiler"]

[dependencies]
rustpython-parser = { path = "../../parser", optional = true }
rustpython-compiler = { path = "../../compiler", optional = true }
rustpython-vm = { path = "../../vm", default-features = false }
```

Then, just add `#[cfg(feature = "parser-compiler")]` attributes to everything in the `rustpython-wasm` crate that uses `rustpython_compiler` or `rustpython_parser` (which should be a lot, because right now we only support passing code as strings), and add new APIs that can accept `Vec<u8>`s and use `rustpython_bytecode::bytecode::CodeObject::from_bytes` to deserialize them.

**Comment by coolreader18** [comment]
You can test if it works by commenting out the `default = [...]` line and compiling the demo (`npm run dist`). Also, I really don't think it's very feasible to try to convert CPython->RustPython bytecode; they're different enough that it would be difficult/inefficient to do so, and it really does not bring much benefit (in my view). Also, it would have to be done at runtime, which would just add to the startup time of the python wasm app.

**Comment by Joetib** [comment]
Thank you very much. I will try the solution out. I have my hopes high already. Off to work!

**Comment by mireille-raad** [comment]
I am also interested in getting python to work in the browser... but the use cases that I know of are related to academia, publishing scientific papers and reproducible research especially in economics.

The load time is not as huge of a hurdle or at least I can manage the user's expectation and some some smart loading of resources.

I'll look the merge request from 10 days ago. I've seen your demo and looked at another repo https://github.com/RustPython/demo where there is a .wasm file and re-used that. Is there an "optimal" way or just a minimal source code to load python in browser? Should we add those to examples?
